diff --git a/sa_yaml/7/validator.php b/sa_yaml/7/validator.php
--- a/sa_yaml/7/validator.php
+++ b/sa_yaml/7/validator.php
@@ -12,6 +12,7 @@
 use Composer\IO\NullIO;
 use Composer\Repository\ComposerRepository;
 use Composer\Repository\RepositoryInterface;
+use Composer\Repository\RepositorySet;
 use Composer\Util\HttpDownloader;
 use Symfony\Component\Console\Application;
 use Symfony\Component\Console\Command\Command;
@@ -301,13 +302,12 @@
         if (!isset($this->composerPackageExists[$key])) {
             $repository = $this->getComposerRepository($repositoryUri);

-            $found = false;
-            foreach ($repository->search($package, RepositoryInterface::SEARCH_NAME) as $result) {
-                if ($result['name'] === $package) {
-                    $found = true;
-                    break;
-                }
-            }
+            $repositorySet = new RepositorySet('dev');
+            $repositorySet->allowInstalledRepositories();
+            $repositorySet->addRepository($repository);
+            $pool = $repositorySet->createPoolForPackage($package);
+            $matches = $pool->whatProvides($package, null);
+            $found = !empty($matches);

             $this->composerPackageExists[$key] = $found;
         }
diff --git a/sa_yaml/8/validator.php b/sa_yaml/8/validator.php
--- a/sa_yaml/8/validator.php
+++ b/sa_yaml/8/validator.php
@@ -12,6 +12,7 @@
 use Composer\IO\NullIO;
 use Composer\Repository\ComposerRepository;
 use Composer\Repository\RepositoryInterface;
+use Composer\Repository\RepositorySet;
 use Composer\Util\HttpDownloader;
 use Symfony\Component\Console\Application;
 use Symfony\Component\Console\Command\Command;
@@ -301,13 +302,12 @@
         if (!isset($this->composerPackageExists[$key])) {
             $repository = $this->getComposerRepository($repositoryUri);

-            $found = false;
-            foreach ($repository->search($package, RepositoryInterface::SEARCH_NAME) as $result) {
-                if ($result['name'] === $package) {
-                    $found = true;
-                    break;
-                }
-            }
+            $repositorySet = new RepositorySet('dev');
+            $repositorySet->allowInstalledRepositories();
+            $repositorySet->addRepository($repository);
+            $pool = $repositorySet->createPoolForPackage($package);
+            $matches = $pool->whatProvides($package, null);
+            $found = !empty($matches);

             $this->composerPackageExists[$key] = $found;
         }
